hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/producers/StreamFileProducerConcurrent.java

```diff
@@ -143,34 +152,40 @@ public void switchBlocks(
 
-        assert lastRecordHashingResult != null : "initRunningHash() must be called before switchBlocks";
-        if (newBlockNumber != lastBlockNumber + 1) {
-            throw new IllegalArgumentException("Block numbers must be sequential, newBlockNumber=" + newBlockNumber
-                    + ", lastBlockNumber=" + lastBlockNumber);
-        }
+        try {
+            assert lastRecordHashingResult != null : "initRunningHash() must be called before switchBlocks";
+            if (newBlockNumber != lastBlockNumber + 1) {
+                throw new IllegalArgumentException("Block numbers must be sequential, newBlockNumber=" + newBlockNumber
+                        + ", lastBlockNumber=" + lastBlockNumber);
+            }
 
-        this.currentBlockNumber = newBlockNumber;
-        requireNonNull(newBlockFirstTransactionConsensusTime);
+            this.currentBlockNumber = newBlockNumber;
+            requireNonNull(newBlockFirstTransactionConsensusTime);
 
-        if (currentRecordFileWriter == null) {
-            // We are at the start of a new block and there is no old one to close or wait for. So just create a new one
-            // which creates a new file and writes initializes it in the background
-            currentRecordFileWriter = lastRecordHashingResult.thenApply(lastRunningHash ->
-                    createBlockRecordWriter(lastRunningHash, newBlockFirstTransactionConsensusTime, newBlockNumber));
-        } else {
-            // wait for all background threads to finish, then in new background task finish the current block
-            currentRecordFileWriter = currentRecordFileWriter
-                    .thenCombine(lastRecordHashingResult, TwoResults::new)
-                    .thenApplyAsync(
-                            twoResults -> {
-                                final var writer = twoResults.a();
-                                final var lastRunningHash = twoResults.b();
-                                closeWriter(writer, lastRunningHash);
-                                return createBlockRecordWriter(
-                                        lastRunningHash, newBlockFirstTransactionConsensusTime, newBlockNumber);
-                            },
-                            executorService);
+            if (currentRecordFileWriter == null) {
+                // We are at the start of a new block and there is no old one to close or wait for. So just create a new
+                // one
+                // which creates a new file and writes initializes it in the background
+                currentRecordFileWriter = lastRecordHashingResult.thenApply(lastRunningHash -> createBlockRecordWriter(
+                        lastRunningHash, newBlockFirstTransactionConsensusTime, newBlockNumber));
+            } else {
+                // wait for all background threads to finish, then in new background task finish the current block
+                currentRecordFileWriter = currentRecordFileWriter
```

Comment:
Worth noting that this won't run the indicated code within the context of the lock; the lastRecordHashingResult used will be whatever value is current when the asynchronous tasks actually run on some other thread.

Reply:
I agree that the future won't run in the context of the lock. My concern was with the setting of the future into `currentRecordFileWriter` as well as the reading of variables like `lastRecordHashingResult`.